Add template for plugin projects.

jakobr-google opened the pull request:
Plugin projects are created by running `flutter create --plugin <name>`.

By default, an example app is also created in the plugin project. This can be disabled by adding `--no-example`. The example app is created using the normal 'create' template, which has been modified to allow for conditional plugin code.

Modified the android package name to match package naming conventions (all lower-case, and must match the directory name).

jakobr-google committed:
Add template for plugin projects.

Plugin projects are created by running `flutter create --plugin <name>`.

By default, an example app is also created in the plugin project. This can be disabled by adding `--no-example`. The example app is created using the normal 'create' template, which has been modified to allow for conditional plugin code.

Modified the android package name to match package naming conventions (all lower-case, and must match the directory name).

devoncarew commented:
Some initial thoughts:

> Plugin projects are created by running flutter create --plugin <name>.

Perhaps use `--plugin-project`, or `--template=plugin`? The `--template=foo` arg would be nice in that `create` would always be using a template, just w/o an explicit arg it would default to the regular project template.

> By default, an example app is also created in the plugin project. This can be disabled by adding --no-example

What about not having a `--no-example` flag? Users can always delete the content, and fewer options is nicer.

collinjackson commented:
Lately I've been putting "Plugin" at the end of the plugin class names on the native side, to make them more distinguishable from the native classes that plugins often wrap, e.g. FirebaseDatabasePlugin vs FirebaseDatabase, ImagePickerPlugin vs ImagePicker

There are other available workarounds such as class prefixing (on iOS) and namespacing (on Android) but having "Plugin" the name makes it more clear what the class does, I think.

collinjackson reviewed (approved):
Also, I kind of agree with Devon that fewer options is better and it's easy for plugin authors to just delete the example if they don't want one.

jakobr-google committed:
Remove --example option. Always generate example app.

jakobr-google committed:
Add Plugin suffix to native classes (if they don't already have it).

jakobr-google commented:
Removed the --example flag, and added a Plugin suffix to the native classes.

I prefer --plugin over --plugin-project since it's shorter, and what else would we be creating? --template=<foo> sounds nice, but there are really two templates involved in creating a plugin, 'plugin' and 'create' (for the example app).

devoncarew commented:
> but there are really two templates involved in creating a plugin, 'plugin' and 'create'

Yes, but it makes things a bit more explicit. Instead of slightly altering the mode for `create`, we're selecting between two options.

Hixie commented on packages/flutter_tools/templates/create/lib/main.dart.tmpl:
```diff
@@ -55,6 +58,19 @@ class MyHomePage extends StatefulWidget {
+    super.initState();
+    {{pluginDartClass}}.platformVersion.then((String platformVersion) {
+      setState(() {
+        _platformVersion = platformVersion;
```
This code is buggy. If the _MyHomePageState state has been disposed before the future completes, it'll assert.

Is there a way to cancel a pending platform request? We should do so in the state's `dispose()` function so that it doesn't unexpectedly complete here in that case.

mit-mit replied:
There is no way to cancel. Added mounted check in https://github.com/flutter/flutter/issues/9458.

Hixie commented on packages/flutter_tools/templates/create/lib/main.dart.tmpl:
```diff
@@ -84,9 +100,21 @@ class _MyHomePageState extends State<MyHomePage> {
+          children: [
+            new Text('Running on: $_platformVersion\n'),
+            new Text(
+                'Button tapped $_counter time${ _counter == 1 ? '' : 's' }.'),
```
this doesn't match the style guide

mit-mit replied:
Fixed in https://github.com/flutter/flutter/issues/9458

Hixie commented on packages/flutter_tools/templates/create/lib/main.dart.tmpl:
```diff
@@ -84,9 +100,21 @@ class _MyHomePageState extends State<MyHomePage> {
       ),
       body: new Center(
+{{#withPluginHook}}
+        child: new Column(
```
If we're going to add Column here we need to add comments explaining what's going on, especially if we add mainAxisSize as well.

I'm worried about having multiple versions of this template. We have four now (normal, driver, plugin, driver+plugin). We have micromanaged the creation of this template based on the usability studies; the more versions we have the more studies we're going to need to actually test them all. This doesn't scale.

mit-mit replied:
I'm not sure what to do about the 'multiple versions' part. We deliberately did the conditional approach, as the alternative would have been to fork ALL ~60 files in /templates/create/ into /templates/plugin/ which would have been a lot less scalable.

Do you have a better idea than the conditional approach?

devoncarew replied:
My 0.02 is that we should have two templates, a normal app project and a plugin project. The normal project should include driver support, example unit tests, ... - things we think are best practice and that we'd like to lead users towards.

mit-mit replied:
We already have that. This is for the example app inside a plugin (the /example/ for)

devoncarew replied:
Ah ok. I thought this was discussing the `flutter create` case, where we do have a matrix of options.

Hixie replied:
What I mean is that right now the main.dart.tmpl file is effectively forked into four versions:
* the one generated by `flutter create`
* the one generated by `flutter create --driver` or some such
* the one generated by `flutter create --plugin` or some such
* the one hypothetically generated by `flutter create --plugin --driver` or some such (I suspect we don't ever generate this today, I haven't checked)

We have carefully crafted the one generated by `flutter create` based on usability feedback to optimise for the first hour experience. The other versions haven't been tested and I'm worried about what will happen as the needs for these various versions diverge.

Hixie replied:
To clarify, I'm not suggesting forking more. That wouldn't address my concerns (it would make in fact them worse, for the reasons discussed above by @mit-mit). My concerns are not concrete, they are merely fear for the future. :-)

mit-mit replied:
Take a look at the PR i just sent. I think the more clear separation in main.dart makes it a little more manageable.
